pkg/source/source.go

```diff
@@ -68,164 +77,249 @@ var _ Source = &Channel{}
+	// Start calls to no-op.
+	canceled bool
+	// destinations are the destination channels of the added event handlers
+	destinations []chan event.GenericEvent
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	group  *errgroup.Group
 }
 
 func (cs *Channel) String() string {
 	return fmt.Sprintf("channel source: %p", cs)
 }
 
-// Start implements Source and should only be called by the Controller.
-func (cs *Channel) Start(
-	ctx context.Context,
-	handler handler.EventHandler,
-	queue workqueue.RateLimitingInterface,
-	prct ...predicate.Predicate) error {
+// Start is internal and should be called only by the Controller to register an EventHandler with the Informer
+// to enqueue reconcile.Requests. It should NOT block, instead, it should start a goroutine and return immediately.
+// The context passed to Start can be used to cancel the blocking operations in the Start method. To cancel the
+// goroutine/ shutdown the source a user should call Stop.
+// Start can be called multiple times, allowing multiple controllers to share a single source.
+func (cs *Channel) Start(ctx context.Context, handler handler.EventHandler, queue workqueue.RateLimitingInterface,
+	predicates ...predicate.Predicate) error {
 	// Source should have been specified by the user.
 	if cs.Source == nil {
 		return fmt.Errorf("must specify Channel.Source")
 	}
 
-	// set the stop channel to be the context.
-	cs.stop = ctx.Done()
-
 	// use default value if DestBufferSize not specified
 	if cs.DestBufferSize == 0 {
 		cs.DestBufferSize = defaultBufferSize
 	}
 
-	dst := make(chan event.GenericEvent, cs.DestBufferSize)
-
-	cs.destLock.Lock()
-	cs.dest = append(cs.dest, dst)
-	cs.destLock.Unlock()
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	if cs.canceled {
+		return nil
+	}
 
-	cs.once.Do(func() {
-		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
-		go cs.syncLoop(ctx)
-	})
+	// Create a destination channel for the event handler
+	// and add it to the list of destinations
+	destination := make(chan event.GenericEvent, cs.DestBufferSize)
+	cs.destinations = append(cs.destinations, destination)
+
+	// Initialize the group and event loop if this is the first time.
+	if cs.group == nil {
+		// Create a new context and group for the event loop.
+		// We want to store the cancellation function and the sub-context
+		// returned from errgroup.WithContext so that we can pass it to the handlers.
+		ctx, cancel := context.WithCancel(ctx)
+		cs.cancel = cancel
+		cs.group, cs.ctx = errgroup.WithContext(ctx)
+
+		// Start the main event loop, given that this is the first time
+		// we are adding an event handler.
+		cs.group.Go(cs.loop)
+	}
 
-	go func() {
-		for evt := range dst {
-			shouldHandle := true
-			for _, p := range prct {
-				if !p.Generic(evt) {
-					shouldHandle = false
-					break
+	// Start a goroutine to handle events for the event handler.
+	cs.group.Go(func() error {
+	eventloop:
+		for {
+			select {
+			case <-ctx.Done():
```

Comment:
```suggestion
			case <-cs.ctx.Done():
```
I think we should use cs.ctx here just like in the main event loop